### ISSUE TITLE
Request: Cache or parallelize preprocessing dataset

### ISSUE THREAD

**Issue by ClimbsRocks** [catboost/catboost#273]
Hi Team,

Great project! We're looking into using it for a bunch of our predictors. Unfortunately, it takes a really long time to train (multiple hours, compared to 5-10 minutes for LightGBM). 

As @annaveronika mentioned [here](https://github.com/szilard/GBM-perf/issues/4#issuecomment-362558773) the feature preprocessing step appears to be quite expensive (even for our large datasets, it can take 10-20% of the total training time). 

For several of our predictors, we use the same input features to predict multiple different outputs (when an order is submitted, we might try to predict how long that order will take, what the impact to retention is, if there are likely to be any issues with that order, etc.). 

**Is there any way to easily cache the results of the preprocessing step**, so we only have to do that once for our suite of related predictors that all use the same input? This would help us train multiple predictors more rapidly. 

Alternatively, **is there any way to parallelize the preprocessing step?** We've got some powerful compute boxes available for model training, and it seems wasteful to have all those cores sitting around unused while waiting for this rather expensive calculation to finish. 

Thanks again for the great project!

**Comment by Noxoomo** [comment]
Hi!

1)  we are planning to add special pool format, which will decrease RAM usage and will be much faster to load.
2) What version do you use: GPU or CPU?
3) What kind of features did you use? CatBoost could run with comparable to LightGBM speed, but you need manually to turn of some advance learning techniques (which are not implemented in LightGBM and allows LightGBM to run faster)

**Comment by annaveronika** [comment]
Also which version do you use? The last one should be comparable in speed with LightGBM.

If you use last version then one hypothesis why it runns slower is that you are using one-hot encoded features. LightGBM has specific optimisation for that. We recommend to use one-hot-max-size parameter and use categorical features directly, then it will run much faster.

Also could you give more details about your setup? How big is the dataset? How much time does preprocessing take in your case? How many iterations of boosting you use?

**Comment by ClimbsRocks** [comment]
Great questions!

1. Faster loading and less RAM usage sounds great. 
2. We're using CPU right now. It's a pretty powerful CPU with a decent number of cores. 
3. Almost entirely numerical features, with just 1 or 2 categorical features that aren't too important. This project is more about accuracy than speed, but if there are one or two advanced features that matter mostly for categorical inputs, or that have relatively little impact on accuracy but an outsized impact on speed, that would be fantastic to know about. 
4. Using version `0.6.3`, installed from PyPi, on Linux, on both Python2 and Python3. We're using the sklearn API in Python.
5. I think I'm feeding in categorical features directly (a single column of categorical values, and passing in a value for cat_features). Should I still specify `one-hot-max-size`?
6. The dataset is a few million rows, and a few hundred or thousand columns. We include an eval_set of a few tends of thousands of rows. Since training was taking so long, I've started drastically bumping down the number of iterations and increasing the learning rate. My current values are something like `{'iterations': 250, 'learning_rate': 0.3}`. Initially we were training for 1000 iterations and a learning rate of 0.05. The preprocessing (time between when we call `cat_boost_model.fit()` and when we see logging for each iteration show up) pretty consistently takes about 20-25 minutes. Ultimately, that's the part I'm hoping to avoid repeating. 

That was one of the things that drastically sped up model training time with LightGBM- they also do some expensive preprocessing of the data. But they offer an option to save the transformed dataset to disk. So we do that once, and then train up a variety of models on different output columns but the same already-processed dataset. 

I'm assuming the preprocessing can either be cached, or perhaps even better, parallelized. Parallelizing it would hopefully bring that time down by an order of magnitude, and would immediately apply to everyone who trains a model. Caching the processed dataset by writing it to disk is probably very useful to only a small handful of superusers.

**Comment by Noxoomo** [comment]
1) Most of preprocessing is already parallelized
2) For  GPU on 56-core Intel Xeon all preprocessing stage for 500GB datasets takes approximately 25-30 minutes to load from disk and start learning.  And with loaded in RAM TPool it should be significantly faster. And I don't think CPU version should take more time, than GPU, to start learning. From you description, dataset you use is significantly smaller, so I suppose there is some issue and let's figure out, where it is:
a) Could you provide more CPU specs (core count, frequency or just model)?
b) What CatBoost parameters do you use?
c) Could you try to learn from command line interface with --logging-level Debug and --detailed-profiel options?  
d) Do you create new TPool for each learning task? 
e) Do you change feature_border_type option? If it's set to MinEntropy (0.6.3 default should be GreedyLogSum), then preprocessing will be slow. 
3) What is cardinality of your categorical features? If it's pretty small (like <= 16-32), then you, probably, could adjust one-hot-max-size and slightly optimize learning

**Comment by annaveronika** [comment]
Firstly, we've made a set of speedups for preprocessing, so it's much faster now. Secondly, preprocessing can now be cached, use quantize method on the Pool object, and then use quantized Pool for your training: https://catboost.ai/docs/concepts/python-reference_pool_quantized.html

**Comment by RiverShah** [comment]
Propose that we reopen the issue. When training with GPUs, pre processing of the data before boosting even begins is roughly 30% of the time. Processor usage is minimal for all but one core during this time while all other host resources sit idle. To test the issue with pre-processing, please try with a large synthetic dataset, for example 100M * 200. We should notice that pre-proc time is high; parallelism is low during this stage. Requesting that we distribute the work to more cores to speed up pre processing time.

While saving the quantized dataset is an option, it is still not ideal and adds extra steps and code complexity and is not suitable for all tasks and model tuning. Thanks for looking into this.

**Comment by RiverShah** [comment]
Propose that we reopen this issue. Please try with a large synthetic numeric dataset with training on GPU (for example Tesla V100). For datasets with shape 100M * 150, with 8 GPU devices per node, I am seeing preprocessing times dominate that of the boosting itself. The vast bulk of the node resources (64 processor cores and 8 GPUs) sit idle for about 5-15 minutes before boosting begins. Can you please test with these large datsets and check if there is anything that can be parallelised or optimised to speed up the pre-boosting phase for large numeric arrays. Thanks!
